pkg/zstdchunked/storage.go

```diff
@@ -0,0 +1,983 @@
+				_, err := io.Copy(ioutil.Discard, limitReader)
+				if err != nil {
+					return err
+				}
+				continue
+			}
+
+			limitReader := io.LimitReader(p, mf.Length())
+
+			if err := createFileFromZstdStream(dest, dirfd, limitReader, missingDirsMode, os.FileMode(mf.File.Mode), mf.File, options); err != nil {
+				p.Close()
+				return err
+			}
+		}
+		p.Close()
+	}
+	return nil
+}
+
+func mergeMissingChunks(missingChunks []missingChunk, target int) []missingChunk {
+	if len(missingChunks) <= target {
+		return missingChunks
+	}
+
+	getGap := func(missingChunks []missingChunk, i int) int {
+		prev := missingChunks[i-1].RawChunk.Offset + missingChunks[i-1].RawChunk.Length
+		return int(missingChunks[i].RawChunk.Offset - prev)
+	}
+
+	// this implementation doesn't account for duplicates, so it could merge
+	// more than necessary to reach the specified target.  Since target itself
+	// is a heuristic value, it doesn't matter.
+	var gaps []int
+	for i := 1; i < len(missingChunks); i++ {
+		gaps = append(gaps, getGap(missingChunks, i))
+	}
+	sort.Ints(gaps)
+
+	toShrink := len(missingChunks) - target
+	targetValue := gaps[toShrink-1]
+
+	newMissingChunks := missingChunks[0:1]
+	for i := 1; i < len(missingChunks); i++ {
+		gap := getGap(missingChunks, i)
+		if gap > targetValue {
+			newMissingChunks = append(newMissingChunks, missingChunks[i])
+		} else {
+			prev := &newMissingChunks[len(newMissingChunks)-1]
+			gapFile := missingFile{
+				Gap: int64(gap),
+			}
+			prev.RawChunk.Length += uint64(gap) + missingChunks[i].RawChunk.Length
+			prev.Files = append(append(prev.Files, gapFile), missingChunks[i].Files...)
+		}
+	}
+
+	return newMissingChunks
+}
+
+func retrieveMissingFiles(input *ChunkedZstdDiffer, dest string, dirfd int, missingChunks []missingChunk, missingDirsMode os.FileMode, options *archive.TarOptions) error {
+	var chunksToRequest []ImageSourceChunk
+	for _, c := range missingChunks {
+		chunksToRequest = append(chunksToRequest, c.RawChunk)
+	}
+
+	// There are some missing files.  Prepare a multirange request for the missing chunks.
+	var stream io.ReadCloser
+	var contentType string
+	var err error
+
+	for stream == nil {
+		stream, contentType, err = input.stream.GetBlobAt(chunksToRequest)
+		if err != nil {
+			if _, ok := err.(ErrBadRequest); ok {
+				requested := len(missingChunks)
+				// If the server cannot handle at least 64 chunks in a single request, just give up.
+				if requested < 64 {
+					return err
+				}
+
+				// Merge more chunks to request
+				missingChunks = mergeMissingChunks(missingChunks, requested/2)
+				continue
+			}
+			return err
+		}
+	}
+
+	if err := storeMissingFiles(stream, contentType, dest, dirfd, missingChunks, missingDirsMode, options); err != nil {
+		stream.Close()
+		return err
+	}
+	if err := stream.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func safeMkdir(target string, dirfd int, mode os.FileMode, metadata *FileMetadata, options *archive.TarOptions) error {
```

Comment:
Maybe it makes sense to make a nice little package out of it (openat2 and friends).

E.g. `github.com/containers/storage/pkg/safe` (or `secure`) with methods like `Open`, `Mkdir` etc.

Reply:
I'll definitely refactor some of this code in a separate package, as some features are not really zstd related